src/lifetimes.md

```diff
@@ -213,3 +213,205 @@ totally ok*, because it keeps us from spending all day explaining our program
+we know that  a `&'s1 str'` outlives an `&'s2 str`, so we can substitute `&'s1
+s1` with `&'s2 s2`. After this both arguments are of lifetime `&'s2` and the
+call-site is consistent with the signature of `print_shortest`.
+
+> More formally, the basis for the above rule is in *type variance*. Under this
+> model, you would consider a longer lifetime a sub-type of a shorter lifetime,
+> and for function arguments to be *co-variant*. However, an understanding of
+> variance isn't strictly required to understand the Rust borrow checker. We've
+> tried here to instead to explain using intuitive terminlolgy.
+
+## Inter-procedural Borrow Checking of Function Return Values
+
+Now consider a slight variation of this example:
+
+```rust,ignore
+fn main() {
+    let s1 = String::from("short");
+    let res;
+    let s2 = String::from("a long long long string");
+    res = shortest(&s1, &s2);
+    println!("{}", res);
+}
+
+fn shortest<'k>(x: &'k str, y: &'k str) -> &'k str {
+    if x.len() < y.len() {
+        return x;
+    } else {
+        return y;
+    }
+}
+```
+
+`print_shortest` has been renamed to `shortest`, which instead of printing,
+now returns the shorter of the two strings. It does this using only references
+for efficiency, avoiding the need to re-allocate a new string to pass back to
+`main`. The responsibility of printing the result has been shifted to `main`.
+
+Let's again de-sugar `main` by adding explicit scopes and lifetimes:
+
+```rust,ignore
+fn main() {
+    's1 {
+        let s1 = String::from("short");
+        'res {
+            let res: &'res str;
+            's2 {
+                let s2 = String::from("a long long long string");
+                res: &'res: str = shortest(&'s1 s1, &'s2 s2);
+                println!("{}", res);
+            }
+        }
+    }
+}
+```
+
+Again, at the call-site of `shortest` the compiler needs to check the
+consistency of the arguments in the caller with the signature of the callee.
+The signature of `shortest` first says that the two reference arguments have
+the same lifetime, which can be prove ok in the same way as before, thus giving
+us:
+
+```rust,ignore
+res: &'res = shortest(&'s2 s1, &'s2 s2);
+```
+
+But we now have the additional reference to check. We must now prove that the
+returned reference can have the same lifetime as the arguments of lifetime
+`&'s2`. This brings us to a second rule:
+
+> The return value of a function `&'r T` can be converted to an argument `&'s T`
+> if the lifetime of `&'r T` is equal or shorter than `&'s T`.
+
+To make our program compile, we would have to subsitute `res: &'res` for `res:
+&'s2`, but we can't since `&'res` in fact out-lives `&'s2`. This program is
+inconsistent and the compiler rightfully rejects the program because we
+try make a reference (`res`) which outlives one of the values it may refer to
+(`s2`).
+
+> Formally, function return values are said to be *contravariant*, the opposite
+> of *covariant*.
```

Comment:
I feel the argument/return position might be a bit of a red herring here.

I wouldn’t use the term contravariant to refer to return types, at least not in this sense.  They appear so because [a return type was compared to an argument type](https://github.com/rust-lang-nursery/nomicon/pull/4/files?diff=unified#diff-d12e10c55aa9b4bcd76634d6be48ab5eR362), but if you compared a return type to another return type they would appear covariant just as well!

---

The lifetime coercion (subtyping) rule is simply `&'a T -> &'b T` provided that `'a` outlives `'b`.  Argument/return type doesn’t play a role here.\*  The compiler is allowed to *shrink* the lifetime of a reference, but never extend it.  Formally speaking, reference types are said to be “lifetime-covariant”.

Another rule is that an object must always outlive its binding.  This means you can’t, in a variable of a larger scope, store a reference to an object of a smaller scope:

```rust
'r_scope {
    let r: &'r_ref;
    'inner {
        let inner = String::new();
        r = &'r_ref inner; // doesn't work because 'r_ref ≥ 'r_scope > 'inner
                           // yet 'r_ref ≤ 'inner
    }
}
```

From the first example, you learned that `'k` has to be `'s2` (or shorter).  This remains true for the second example.

In the second example, the returned reference has lifetime `'k`.  When it gets stored in `res`, the compiler may shrink `'k` if necessary.  Therefore the lifetime `'res_ref` of the stored reference must be `'k` or shorter.  Here, I used `'res_ref` to distinguish between the lifetime of the reference `res: &'res_ref str`, and the lifetime of the scope itself `'res_scope { let res; ... }`.

From the structure of the code itself, we know that the scope of `s2` is contained fully within that of `res`.  Because an object must always outlive its binding, it follows that `'res_ref` must outlive `'res_scope`, which is strictly longer than `'s2`.

If we combine these three facts together,

  - `'k` must be `'s2` or shorter
  - `'res_ref` must be `'k` or shorter
  - `'res_ref` must be strictly longer than `'s2`

we see that there can be no solution that satisfies all three constraints :(

There is a way to fix it though, by making `'s2` and `'res_scope` equal:

```rust
fn main() {
    let s1 = String::from("short");
    let (res, s2);
    s2 = String::from("a long long long string");
    res = shortest(&s1, &s2);
    println!("{}", res);
}
```

--

\* Due to variance, the rule changes for more complicated for composite types like `Cell<&'a ()>` or `fn(&'a ()) -> &'b ()` but that doesn't apply here.

Reply:
@vext01 any chance you can find the time sometime soon?

Reply:
I'll try to find time this week.

I've learnt more Rust since I last looked at this, so I may have some edits of my own.